### ISSUE TITLE
Implement onion nodes' public key retrieval in directory node

### ISSUE THREAD

**Issue by DestructiveReasoning** [CamiloGarciaLaRotta/Cebolla#25]
The directory node needs to get the public keys from the running onion nodes.
I'll outline a brief list of what this entails.

- [x] Firstly, the onion routers need to *have* key pairs. This can be done by assigning an `RSAVirtuoso` to the onion routers, as follows:
```Python
import sys
sys.path.append('../stealth')
import stealth

keypair = RSAVirtuoso()
```
- [x] Also, the directory node needs its own keypair. This can of course be implemented the same way. 
- [x] After the directory node figures out which onion nodes are running, it should make a TCP connection with each one in order to receive their public keys.
- [x] Here's where things might get weird. Currently, in the onion router's main function, they immediately listen for connections that are interpreted as establishments. But now, connections may be made for the purpose of getting their public key, which is completely separate from onion routing. I propose two ways of doing this: firstly, the onion routers may interpret their first connection ever as a public key request, and therefore once it accepts a connection it immediately sends its public key back.  The problem with this is that if a directory node shuts down (or if another directory node starts up) after it has already given its public key, then it will never give its public key again. The alternate solution is to have the onion routers run another thread listening on another port, and interprets all messages to that port as public key requests. This is slightly more complex. Regardless of how its done, here's how the onion router should get its public key:
```Python
key = keypair.get_public_key().exportKey()
```
Keep in mind, this returns a byte string, so use `.decode('utf-8')` if you need a string. Don't base64 encode/decode it.
- [x] Once the directory node receives all public keys, use those instead of the dummies :)

**Comment by stumash** [comment]
I like the thread option.  so the routers have a special thread that listens on a special port for key requests.

but do you as please.

**Comment by DestructiveReasoning** [comment]
@stumash 
Yeah, the thread option is definitely more robust. That being said, for the purpose of the project I'm sure the drawback I pointed out for my other proposal is probably not too big of a deal. It might be worth it for its simplicity. But I'll let @CamiloGarciaLaRotta decide how he wants to implement it.

**Comment by CamiloGarciaLaRotta** [comment]
@DestructiveReasoning @stumash 

**Problem**
Crypto library is only installed for Python2.7 in the CS machines. Python3 does not have it.
We don't have install rights in the machines.

**Attempted solutions**
- Pull and manually install library -> failed, still requires installing modules in /urs/local/python3.
- Copy ``/usr/local/lib/python2.7/dist-packages/Crypto`` into Cebolla and import it into ``stealth.py`` -> Crypto dependencies are attached to Python2.7, fails to import all submodules 
- Use porting tools like 2to3's module ``from past import ...`` -> 2to3 modules aren't installed either.
- The last attempt to avoid having to recode everything with either another crypto library for python3 or everything in python2: emailed the CS admins to ask to install the library on Python3. I had asked over the summer to install the Heroku SDK and they did it without any questions. Hopefully, it will be the same now.

**Comment by CamiloGarciaLaRotta** [comment]
Awaiting reply from CS sysadmins ported py3 to py2 code, fully tested 9fa14493126c2ac7f40ff1c33ed2862bbd7650e0

**Comment by CamiloGarciaLaRotta** [comment]
Fixed with #26
